### ISSUE TITLE
Automation with Travis

### ISSUE THREAD

**Issue by reiner-dolp** [liblouis/js-build#2]
[removed]

**Comment by bertfrees** [comment]
Are you planning to tag the js-build commits manually or should they be based on the liblouis tags?

It looks like a command is missing to build build-tables-embeded.js.

**Comment by reiner-dolp** [comment]
Its the same command with an additional `--embed-file` flag. I added the command above.

Builds of tagged commits/official releases in the main repo should have the same tag in this repo: `v3.1.0`, `v3.0.0`, ...

I would love to have builts of all commits in `liblouis/liblouis` that change the output/behaviour of liblouis. (Which should be commits that change a file in the `tables` folder or affect the resulting `liblouis.so` file.) However, we need a naming scheme that makes it easy to map commits in this repo to commits in the main repo `liblouis/liblouis` and at the same time, generates a correct version ordering in package managers. That's why issue #1 exists.

**Comment by bertfrees** [comment]
It is probably easiest to build a "snapshot" for every single commit on master. Otherwise you would have to determine somehow when there is a behavioral change which doesn't sound trivial.

**Comment by reiner-dolp** [comment]
The example `travis.yml` I linked above does the following to detect commits that only update documentation:

```yml
# If there are no changes to the compiled out (e.g. this is a README update) then just bail.
cd out
if [ -z `git diff --exit-code` ]; then
    echo "No changes to the output on this push; exiting."
    exit 0
fi
```
However, pushing every snapshot would be ok, too.

**Comment by reiner-dolp** [comment]
@bertfrees where in the main repo should I place the shell-scripts for travis ci?

**Comment by bertfrees** [comment]
Inside a folder named ".travis" maybe?

Could the following lines be removed, because Travis has already done this?

```sh
git clone https://github.com/liblouis/liblouis.git
cd liblouis
```

Also I would like to move some parts, notably the `emconfigure`, `emmake` and `emcc` lines, to the Makefile, but that can wait until later. (The `pre.js`, `post.js` and `append.js` files would have to be moved too.)

**Comment by egli** [comment]
@reiner-dolp I'd call the directory just `scripts` (maybe with a `README` inside explaining its purpose), not sure if a hidden directory is such a great idea

**Comment by bertfrees** [comment]
I thought of "travis" because the script is Travis specific, and hidden because ".travis.yml" is also hidden. It doesn't need to be hidden but it makes sense because you'll never run it manually.

**Comment by egli** [comment]
Ok, @bertfrees's arguments make sense

**Comment by bertfrees** [comment]
It may be worth thinking about caching some stuff such as the emscripten dependency.

**Comment by bertfrees** [comment]
```
git clone https://github.com/juj/emsdk.git
```
```
git clone https://github.com/liblouis/liblouis-js.git
```
```
git clone https://github.com/liblouis/js-build.git
```

It's slightly faster to do shallow clones: `git clone --depth 1 ...`

```
git tag -a ${VERSION}-rc
```

```
npm version ${TRAVIS_TAG}-rc
```

Is it the intention that these are different?

**Comment by reiner-dolp** [comment]
I created a directory structure as follows:
```
.travis/
    before_install/
        gcc.sh
        mingw.sh
        emscripten.sh
    script/
        gcc.sh
        mingw.sh
        emscripten.sh
```
Structure of `.travis.yml`:
```yml
sudo: required
dist: trusty

language: c

matrix:
  include:
    - compiler: gcc
      env: BUILD_GCC=yes
    - compiler: gcc
      env: BUILD_GCC=yes ENABLE_UCS4=--enable-ucs4
    - compiler: i586-mingw32msvc-gcc
      env: BUILD_MINGW=yes
    - compiler: gcc
      env: BUILD_JS=yes
      addons: {apt: {sources: [ubuntu-toolchain-r-test], packages: [cmake, g++-4.9]}}

before_install:
  - sudo apt-get update -qq
  - 'chmod +x ./.travis/before_install/*.sh'
  - 'chmod +x ./.travis/script/*.sh'
  - if [ "$BUILD_MINGW" == yes ]; then
        ./.travis/before_install/mingw.sh;
    elif [ "$BUILD_JS" == yes ]; then
        ./.travis/before_install/emscripten.sh;
    else
        ./.travis/before_install/gcc.sh;
    fi

script:
  - cd $TRAVIS_BUILD_DIR
  - if [ "$BUILD_MINGW" == yes ]; then
        ./.travis/script/mingw.sh;
    elif [ "$BUILD_JS" == yes ]; then
        ./.travis/script/emscripten.sh;
    else
        ./.travis/script/gcc.sh;
    fi

after_failure:
  - if [ -z "$BUILD_JS" ]; then
        cat tests/test-suite.log;
    fi
```

**Comment by bertfrees** [comment]
Nice!

**Comment by reiner-dolp** [comment]
I am currently running into the issue that make errors on:
```
No rule to make target `liblouis-12.def', needed by `all-am'.  Stop.
```
I had this issue before on an arch linux machine. Do you know what I am doing wrong? Should `.def` files be built on linux at all?

**Comment by bertfrees** [comment]
You're cross-compiling for Windows?

**Comment by reiner-dolp** [comment]
with the `emcc` commands above to js. [Full Travis Log.](https://travis-ci.org/reiner-dolp/liblouis-fork-detached/jobs/215843970)

**Comment by bertfrees** [comment]
By the way you should use `set -e`, or add `&&` between the lines in the script.

**Comment by bertfrees** [comment]
If you run the emcc command locally, do you see lines like these?
> /bin/bash ../libtool  --tag=CC   --mode=link /home/travis/build/reiner-dolp/liblouis-fork-detached/emscripten-sdk/emscripten/incoming/emcc   -version-info 13:0:1 -no-undefined  -Wl,--output-def,**liblouis-12.def**  -o liblouis.la -rpath /usr/local/lib compileTranslationTable.lo lou_translateString.lo lou_backTranslateString.lo commonTranslationFunctions.lo findTable.lo pattern.lo logging.lo wrappers.lo  ../gnulib/libgnu.la

**Comment by egli** [comment]
`Makefile.am` contains [some code](https://github.com/liblouis/liblouis/blob/master/liblouis/Makefile.am#L25) to generate the def file. I did this along the [documentation](https://www.gnu.org/software/gnulib/manual/html_node/Visual-Studio-Compatibility.html) for the [lib-msvc-compat](http://git.savannah.gnu.org/gitweb/?p=gnulib.git;a=blob_plain;f=modules/lib-msvc-compat) gnulib module. That might not work properly. I looked for `checking if gcc/ld supports -Wl,--output-def... yes` in the [travis log](https://travis-ci.org/reiner-dolp/liblouis-fork-detached/jobs/215843970#L641) and it claims that the linker supports it.

Maybe under some circumstances you are cross compiling but using a different linker that configure thinks?

**Comment by bertfrees** [comment]
> Maybe under some circumstances you are cross compiling but using a different linker that configure thinks?

Might be worth looking at the environment variables that Travis sets. I ran into the problem that Travis was setting `CC` which made libtool use the wrong gcc even though I requested i586-mingw32msvc-gcc (with `--host i586-mingw32msvc`). Maybe related?

**Comment by reiner-dolp** [comment]
This issue is way harder to resolve than I suspected. I was able to reproduce the issue locally by running a Travis CI instance of `dist: trusty` in a docker container. However, I could not figure out why `HAVE_LD_OUTPUT_DEF` is always set (, which is the reason for the `No rule to make target 'liblouis-12.def'` error.)

I discovered that there are small [unofficial docker images of emscripten](https://github.com/asRIA/emscripten-docker). Building using these docker images on my machine works great. But they generate the same error `No rule to make target 'liblouis-12.def'` when run on Travis CI. I do not understand that at all...

I will stop working on this for some days. I will archive the code in a pull request to the main repository.
